proxylib/memcached/text/parser.go

```diff
@@ -0,0 +1,468 @@
+			log.Error("Could not parse text memcache frame")
+			return proxylib.ERROR, 0
+		}
+		logEntry := &cilium.LogEntry_GenericL7{
+			&cilium.L7LogEntry{
+				Proto: "textmemcached",
+				Fields: map[string]string{
+					"command": string(meta.command),
+					"keys":    string(bytes.Join(meta.keys, []byte(", "))),
+				},
+			},
+		}
+
+		p.requestCount++
+		r := &replyIntent{
+			requestID: p.requestCount,
+			command:   meta.command,
+		}
+
+		matches := p.connection.Matches(meta)
+
+		if matches {
+			r.denied = false
+			p.replyQueue = append(p.replyQueue, r)
+			p.connection.Log(cilium.EntryType_Request, logEntry)
+			return proxylib.PASS, frameLength
+		}
+
+		r.denied = true
+		if !noreply {
+			p.injectDeniedMessage()
+			if p.requestCount == p.replyCount+1 {
+			} else {
+				p.replyQueue = append(p.replyQueue, r)
+			}
+		}
+		p.connection.Log(cilium.EntryType_Denied, logEntry)
+		return proxylib.DROP, frameLength
+	} else {
+		//reply
+		log.Debugf("reply, parsing to figure out if we have it all")
+
+		intent := p.replyQueue[0]
+
+		logEntry := &cilium.LogEntry_GenericL7{
+			&cilium.L7LogEntry{
+				Proto: "textmemcached",
+				Fields: map[string]string{
+					"command": string(intent.command),
+				},
+			},
+		}
+
+		if p.isErrorReply(tokens[0]) ||
+			p.isCommandStorage(intent.command) ||
+			p.isCommandDelete(intent.command) ||
+			p.isCommandIncrDecr(intent.command) ||
+			bytes.Equal(intent.command, []byte("touch")) ||
+			bytes.Equal(intent.command, []byte("slabs")) ||
+			bytes.Equal(intent.command, []byte("lru")) ||
+			bytes.Equal(intent.command, []byte("flush_all")) ||
+			bytes.Equal(intent.command, []byte("cache_memlimit")) ||
+			bytes.Equal(intent.command, []byte("version")) ||
+			bytes.Equal(intent.command, []byte("misbehave")) {
+
+			// passing one line of reply
+			p.connection.Log(cilium.EntryType_Response, logEntry)
+			return proxylib.PASS, linefeed + 2
+		} else if p.isCommandRetrieval(intent.command) ||
+			bytes.Equal(intent.command, []byte("stats")) {
+			t, nBytes := p.untilEnd(data)
+			if t == proxylib.PASS {
+				p.connection.Log(cilium.EntryType_Response, logEntry)
+				p.replyQueue = p.replyQueue[1:]
+			}
+			return t, nBytes
+		} else if bytes.Equal(intent.command, []byte("lru_crawler")) {
+			// check if it's response line
+			if bytes.Equal(tokens[0], []byte("OK")) ||
+				bytes.Equal(tokens[0], []byte("BUSY")) ||
+				bytes.Equal(tokens[0], []byte("BADCLASS")) {
+				p.connection.Log(cilium.EntryType_Response, logEntry)
+				return proxylib.PASS, linefeed + 2
+			}
+
+			t, nBytes := p.untilEnd(data)
+			if t == proxylib.PASS {
+				p.connection.Log(cilium.EntryType_Response, logEntry)
+				p.replyQueue = p.replyQueue[1:]
+			}
+			return t, nBytes
+		}
+		log.Error("Could not parse text memcache frame")
+		return proxylib.ERROR, 0
+	}
+}
+
+func (p *TextMemcacheParser) untilEnd(data []byte) (proxylib.OpType, int) {
+	// TODO: optimise this to not ask per byte, but take VALUES lines into account
```

Comment:
create GH issue for this TODO

Reply:
https://github.com/cilium/cilium/issues/5772